Sources/AblyAssetTrackingSubscriber/WorkerQueue/SubscriberWorkerQueueProperties.swift

```diff
@@ -1,7 +1,181 @@
+                    presentPublisherMemberKeys.remove(presenceMessage.memberKey)
+                } else if (presenceMessage.action == .present || presenceMessage.action == .enter || presenceMessage.action == .update){
+                    presentPublisherMemberKeys.insert(presenceMessage.memberKey)
+                    
+                    if let publisherResolution = presenceMessage.data.resolution {
+                        pendingPublisherResolutions.add(resolution: publisherResolution)
+                    }
+                    
+                }
+            }
+        }
+        delegateStateEventsIfRequired()
+    }
+    
+    mutating func delegateStateEventsIfRequired() {
+        let isAPublisherPresent = !presentPublisherMemberKeys.isEmpty
+        
+        var trackableState: ConnectionState?
+        switch lastConnectionStateChange.state {
+        case .online:
+            switch lastChannelConnectionStateChange.state {
+            case .online:
+                trackableState = isAPublisherPresent ? .online : .offline
+            case .offline:
+                trackableState = .offline
+            case .failed:
+                trackableState = .failed
+            case .publishing:
+                break
+            }
+            break
+        case .publishing:
+            break
+        case .offline:
+            trackableState = .offline
+        case .failed:
+            trackableState = .failed
+        }
+        
+        if let trackableState = trackableState, trackableState != lastEmittedTrackableState {
+            lastEmittedTrackableState = trackableState
+            notifyTrackableStateUpdated(trackableState: trackableState)
+        }
+        // It is possible for presentPublisherMemberKeys to not be empty, even when there's no connectivity from our side,
+        // because it's possible to have presence entry events without subsequent leave events.
+        // Therefore, from the perspective of a user consuming events from publisherPresenceStateFlow, what matters
+        // is what's computed for isPublisherVisible (not the simple isAPublisherPresent).
+        var isPublisherVisible = false
+        if case .online = lastConnectionStateChange.state {
+            if isAPublisherPresent {
+                isPublisherVisible = true
+            }
+        }
+        
+        if lastEmittedValueOfIsPublisherVisible != isPublisherVisible {
+            lastEmittedValueOfIsPublisherVisible = isPublisherVisible
+            notifyPublisherPresenceUpdated(isPublisherPresent: isPublisherVisible)
+        }
+        
+        notifyResolutionsChanged(resolutions: pendingPublisherResolutions.drain())
+    }
+    
+    mutating func notifyEnhancedLocationUpdated(locationUpdate: LocationUpdate) {
+        enhancedLocation = locationUpdate
+        delegate?.subscriber(sender: ablySubscriber, didUpdateEnhancedLocation: locationUpdate)
```

Comment:
I'd suggest that these methods that need to perform delegate-related logging should log to a parameter `logHandler: InternalLogHandler?` so that these messages get logged in the context of a particular worker. So their API would look something like [this](https://github.com/ably/ably-asset-tracking-swift/commit/f5566b61f35247aea9b9ed0ffbd46c9e710ee141) (just a temporary commit on my branch; would be best if it were done in this PR) and workers would use them like [this](https://github.com/ably/ably-asset-tracking-swift/commit/2aa095269f078af0620c82683ad0ff7f24149ae1#diff-a2ad5f62ece72ca1721999eeb948e58abd7dbd613048747077f8417a16d714beR16-R19).

Reply:
It seems you've addressed the logging but not the dispatching to the main thread; please can you add that too.

I’d suggest keeping all of this logging / dispatching logic in a single place, same as the `DefaultSubscriber.callback(event: DelegateEvent)` method I linked to above. Easier to keep it consistent that way, what do you think?

Reply:
46713fbe215964c99ad417d7b83e3f64b993602c